sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -59,6 +59,30 @@ trait PredicateHelper {
     }
   }
 
+  /**
+   * Convert an expression into its conjunctive normal form (CNF), i.e. AND of ORs.
+   * For example, a && b || c is normalized to (a || c) && (b || c) by this method.
+   *
+   * Refer to https://en.wikipedia.org/wiki/Conjunctive_normal_form for more information
+   */
+  protected def cnfNormalization(condition: Expression): Expression = {
+    condition.transformUp {
+      case or: Or => pushOrToBottom(or)
+    }
+  }
+
+  private def pushOrToBottom(condition: Expression): Expression = {
```

Comment:
Maybe move this into `cnfNormalization` as a nested function and just name it as `normalize`.